go/libraries/doltcore/schema/index_coll.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import "fmt"
+
+//TODO: column renames need to change colToTag
+//TODO: tag changes need to do...a lot
+type IndexCollection interface {
+	// AddIndexCol adds an index with the given name, columns (in index order), and table primary keys (also in order).
+	// Ex. For table {k1, k2, k3} PK(k1, k2) IDX(k3, k2), columns should be (k3, k2) and pks (k1, k2)
+	AddIndexCol(indexName string, cols []string, pks []string) (Index, error)
+	// AddIndexTag adds an index with the given name, column allTags (in index order), and table pk allTags (also in order).
+	AddIndexTag(indexName string, tags []uint64, pks []uint64) (Index, error)
+	// AllIndexes returns a slice containing all of the indexes in this collection.
+	AllIndexes() []Index
+	// BestMatch takes a collection of columns and finds the index that best matches, if one exists.
+	BestMatch(cols ...string) Index
```

Comment:
What's this for? The engine should be in charge of matching expressions to indexes